core/src/main/java/com/linecorp/armeria/common/AggregatedHttpObject.java

```diff
@@ -36,9 +36,17 @@ public interface AggregatedHttpObject {
     HttpData content();
 
     /**
-     * Returns the content of this message as a string encoded in the specified {@link Charset}.
+     * Returns the content decoded in the {@link Charset} of the content-type header if present,
+     * or the content decoded in the specified {@link Charset} otherwise.
      */
     default String content(Charset charset) {
+        if (headers().contentType() != null && headers().contentType().charset() != null) {
+            Charset contentTypeCharset = headers().contentType().charset();
+            if (!charset.equals(contentTypeCharset)) {
+                return content().toString(contentTypeCharset);
+            }
```

Comment:
To simplify this, you could always use `contentTypeCharset` when it's not `null`, i.e. no need to compare two charsets.